arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/ErrorHandlers.kt

```diff
@@ -0,0 +1,111 @@
+@file:OptIn(ExperimentalTypeInference::class)
+
+package arrow.core.continuations
+
+import arrow.core.nonFatalOrThrow
+import kotlin.experimental.ExperimentalTypeInference
+import kotlin.jvm.JvmMultifileClass
+import kotlin.jvm.JvmName
+
+/**
+ * Catch the shifted value [E] of the `Effect`.
+ * You can either return a value a new value of [A],
+ * or short-circuit the effect by shifting with a value of [E],
+ * or raise an exception into [suspend].
+ *
+ * ```kotlin
+ * import arrow.core.continuations.effect
+ * import arrow.core.continuations.catch
+ *
+ * object User
+ * object Error
+ *
+ * val error = effect<Error, User> { shift(Error) } // // Shift(error)
+ *
+ * val a = error.catch<Error, Error, User> { error -> User } // Success(User)
+ * val b = error.catch<Error, String, User> { error -> shift("other-failure") } // Shift(other-failure)
+ * val c = error.catch<Error, Nothing, User> { error -> throw RuntimeException("BOOM") } // Exception(BOOM)
+ * ```
+ * <!--- KNIT example-effect-error-01.kt -->
+ */
+public infix fun <E, E2, A> Effect<E, A>.catch(@BuilderInference resolve: suspend Shift<E2>.(shifted: E) -> A): Effect<E2, A> =
+  effect { catch(resolve) }
+
+/**
+ * Attempt to run the effect, and [recover] from any unexpected exceptions.
+ * You can either return a value a new value of [A],
+ * or short-circuit the effect by shifting with a value of [E],
+ * or raise an exception into [suspend].
+ *
+ * ```kotlin
+ * import arrow.core.continuations.effect
+ * import arrow.core.continuations.attempt
+ *
+ * object User
+ * object Error
+ *
+ * val exception = effect<Error, User> { throw RuntimeException("BOOM") }  // Exception(BOOM)
+ *
+ * val a = exception.attempt { error -> error.message?.length ?: -1 } // Success(5)
+ * val b = exception.attempt { shift(Error) } // Shift(error)
+ * val c = exception.attempt { throw  RuntimeException("other-failure") } // Exception(other-failure)
+ * ```
+ * <!--- KNIT example-effect-error-02.kt -->
+ */
+public infix fun <E, A> Effect<E, A>.attempt(@BuilderInference recover: suspend Shift<E>.(throwable: Throwable) -> A): Effect<E, A> =
+  effect { attempt(recover) }
+
+/**
+ * A version of [attempt] that refines the [Throwable] to [T].
+ * This is useful for wrapping foreign code, such as database, network calls, etc.
+ *
+ * ```kotlin
+ * import arrow.core.continuations.effect
+ * import arrow.core.continuations.attempt
+ *
+ * object User
+ * object Error
+ *
+ * val x = effect<Error, User> {
+ *   throw IllegalArgumentException("builder missed args")
+ * }.attempt { shift(Error) }
+ * ```
+ *
+ * If you don't need an `error` value when wrapping your foreign code you can use `Nothing` to fill the type parameter.
+ *
+ * ```kotlin
+ * val y = effect<Nothing, User> {
+ *   throw IllegalArgumentException("builder missed args")
+ * }.attempt<IllegalArgumentException, Error, User> { shift(Error) }
+ * ```
+ * <!--- KNIT example-effect-error-03.kt -->
+ */
+@JvmName("attemptOrThrow")
```

Comment:
nit: it does not really `throw` since it captures what it throws back into an effect when it does not match `T`.
`orThrow` implies when I read it that it is unsafe, and that a call when it fails blows up with an exception. Perhaps we can use something like `attemptWithThrowableType` or a similar JVM name indicating we are just matching on type args and letting others unmatched go through but this method is actually safe.

Reply:
Right... I think I originally had this on the DSL method, where it made more sense since it threw immediately into the `suspend` context. It's only needed when you call this API from Java, so it's not super critical.

`attemptRefined`, `attempt`, or maybe just `attemptReified` since it also would have a `Class<T>` param in the Java API?

For reference from Scala APIs:
 -   Cats-effect: `def rethrow[F : MonadError[F, E], A, EE <: E](fa: F[Either[EE, A]]): F[A]`
 - ZIO: `def refineOrDie[E1](pf: PartialFunction[E, E1])(implicit ev1: <:<[E, Throwable], ev2: CanFail[E]): ZIO[R, E1, A]`

Reply:
I just realized that this might not matter since you can't call inline reified methods from Java unless you provide an additional wrapper function that takes a java `Class` as an argument. In any case, a better name would be, as you said, `attemptReified` or similar.